File: app/src/main/java/com/example/soccergamemanager/domain/LineupGenerator.kt
package com.example.soccergamemanager.domain

class LineupGenerator {
    fun generate(
        template: GameTemplateConfig,
        players: List<LineupPlayer>,
        historyByPlayer: Map<String, PlayerSeasonHistory>,
        manualGroupLocks: List<ManualGroupLock> = emptyList(),
    ): LineupGenerationResult {
        val warnings = mutableListOf<String>()
        if (players.size < template.positions.size) {
            return LineupGenerationResult(
                assignments = emptyList(),
                warnings = listOf("At least ${template.positions.size} available players are needed to fill every position."),
            )
        }

        val playerById = players.associateBy { it.id }
        val sanitizedLocks = sanitizeManualLocks(
            template = template,
            manualGroupLocks = manualGroupLocks,
            playerById = playerById,
        )
        warnings += sanitizedLocks.warnings

        val halfKeeperSelections = selectKeepers(
            players = players,
            halfCount = template.halfCount,
            historyByPlayer = historyByPlayer,
            manualLocksByHalfGroup = sanitizedLocks.locksByHalfGroup,
        )
        val firstHalfGroups = mutableMapOf<String, PositionGroup>()
        val assignments = mutableListOf<GeneratedAssignment>()

        repeat(template.halfCount) { halfOffset ->
            val halfNumber = halfOffset + 1
            val keeper = halfKeeperSelections[halfOffset]
            val locksForHalf = sanitizedLocks.locksByHalfGroup[halfNumber].orEmpty()
            val lockedFieldPlayersByGroup = listOf(
                PositionGroup.DEFENSE,
                PositionGroup.LR_MID,
                PositionGroup.CM_STRIKER,
            ).associateWith { group ->
                locksForHalf[group].orEmpty().filterNot { it.id == keeper.id }
            }
            val fieldPlayers = players.filterNot { it.id == keeper.id }
            val capacities = createGroupCapacities(
                fieldPlayerCount = fieldPlayers.size,
                lockedCounts = lockedFieldPlayersByGroup.mapValues { it.value.size },
            )
            if (lockedFieldPlayersByGroup.values.sumOf { it.size } > fieldPlayers.size) {
                warnings += "Half $halfNumber has more locked field players than available field spots."
            }
            if (capacities.values.any { it < 3 }) {
                warnings += "Half $halfNumber cannot keep all field groups at three players with the current availability."
            }

            val priorHalfGroups = if (halfNumber == 2) firstHalfGroups else emptyMap()
            if (halfNumber == 2) {
                lockedFieldPlayersByGroup.forEach { (group, groupedPlayers) ->
                    groupedPlayers.forEach { player ->
                        if (priorHalfGroups[player.id] == group) {
                            warnings += "${player.name} is manually locked into ${group.label} for both halves."
                        }
                    }
                }
            }
            val groupedPlayers = assignFieldGroups(
                fieldPlayers = fieldPlayers,
                capacities = capacities,
                lockedPlayersByGroup = lockedFieldPlayersByGroup,
                priorHalfGroups = priorHalfGroups,
                historyByPlayer = historyByPlayer,
            )

            if (halfNumber == 1) {
                groupedPlayers.forEach { (group, grouped) ->
                    grouped.forEach { player -> firstHalfGroups[player.id] = group }
                }
            }

            for (roundIndex in 1..template.roundsPerHalf) {
                assignments += GeneratedAssignment(
                    halfNumber = halfNumber,
                    roundIndex = roundIndex,
                    playerId = keeper.id,
                    position = FieldPosition.GOALIE,
                    positionGroup = PositionGroup.GOALIE,
                )
            }

            listOf(
                PositionGroup.DEFENSE to listOf(FieldPosition.LEFT_DEFENSE, FieldPosition.RIGHT_DEFENSE),
                PositionGroup.LR_MID to listOf(FieldPosition.LEFT_MIDFIELDER, FieldPosition.RIGHT_MIDFIELDER),
                PositionGroup.CM_STRIKER to listOf(FieldPosition.CENTER_MIDFIELDER, FieldPosition.STRIKER),
            ).forEach { (group, positions) ->
                val groupAssignments = generateGroupAssignments(
                    halfNumber = halfNumber,
                    roundsPerHalf = template.roundsPerHalf,
                    positions = positions,
                    players = groupedPlayers[group].orEmpty(),
                    historyByPlayer = historyByPlayer,
                )
                assignments += groupAssignments
            }
        }

        return LineupGenerationResult(
            assignments = assignments.sortedWith(
                compareBy(
                    GeneratedAssignment::halfNumber,
                    GeneratedAssignment::roundIndex,
                    { template.positions.indexOf(it.position) },
                ),
            ),
            warnings = warnings.distinct(),
        )
    }

    private fun selectKeepers(
        players: List<LineupPlayer>,
        halfCount: Int,
        historyByPlayer: Map<String, PlayerSeasonHistory>,
        manualLocksByHalfGroup: Map<Int, Map<PositionGroup, List<LineupPlayer>>>,
    ): List<LineupPlayer> {
        val keepers = mutableListOf<LineupPlayer>()
        repeat(halfCount) {
            val halfNumber = it + 1
            val lockedKeeper = manualLocksByHalfGroup[halfNumber]?.get(PositionGroup.GOALIE)?.firstOrNull()
            if (lockedKeeper != null) {
                keepers += lockedKeeper
                return@repeat
            }
            val pool = players
                .filter { candidate ->
                    keepers.none { it.id == candidate.id } || players.size < halfCount
                }
                .ifEmpty { players }

            val keeper = pool.minWithOrNull(
                compareBy<LineupPlayer>(
                    { if (it.preferredKeeper) 0 else 1 },
                    { historyByPlayer[it.id]?.keeperAssignments ?: 0 },
                    { historyByPlayer[it.id]?.minutesPlayed ?: 0.0 },
                    { it.name },
                ),
            ) ?: players.first()
            keepers += keeper
        }
        return keepers
    }

    private fun createGroupCapacities(
        fieldPlayerCount: Int,
        lockedCounts: Map<PositionGroup, Int>,
    ): Map<PositionGroup, Int> {
        val fieldGroups = listOf(PositionGroup.DEFENSE, PositionGroup.LR_MID, PositionGroup.CM_STRIKER)
        val base = fieldPlayerCount / fieldGroups.size
        val extra = fieldPlayerCount % fieldGroups.size
        val capacities = fieldGroups
            .associateWith { group ->
                val index = fieldGroups.indexOf(group)
                base + if (index < extra) 1 else 0
            }
            .toMutableMap()

        fieldGroups.forEach { group ->
            val lockedCount = lockedCounts.getOrDefault(group, 0)
            var deficit = lockedCount - capacities.getOrDefault(group, 0)
            if (deficit <= 0) return@forEach

            capacities[group] = lockedCount
            val donorGroups = fieldGroups
                .filter { it != group }
                .sortedByDescending { capacities.getOrDefault(it, 0) - lockedCounts.getOrDefault(it, 0) }

            donorGroups.forEach { donor ->
                while (deficit > 0 && capacities.getOrDefault(donor, 0) > lockedCounts.getOrDefault(donor, 0)) {
                    capacities[donor] = capacities.getOrDefault(donor, 0) - 1
                    deficit -= 1
                }
            }
        }

        return capacities
    }

    private fun assignFieldGroups(
        fieldPlayers: List<LineupPlayer>,
        capacities: Map<PositionGroup, Int>,
        lockedPlayersByGroup: Map<PositionGroup, List<LineupPlayer>>,
        priorHalfGroups: Map<String, PositionGroup>,
        historyByPlayer: Map<String, PlayerSeasonHistory>,
    ): Map<PositionGroup, List<LineupPlayer>> {
        val result = mutableMapOf(
            PositionGroup.DEFENSE to lockedPlayersByGroup[PositionGroup.DEFENSE].orEmpty().toMutableList(),
            PositionGroup.LR_MID to lockedPlayersByGroup[PositionGroup.LR_MID].orEmpty().toMutableList(),
            PositionGroup.CM_STRIKER to lockedPlayersByGroup[PositionGroup.CM_STRIKER].orEmpty().toMutableList(),
        )
        val remaining = capacities
            .mapValues { (group, capacity) -> capacity - result.getValue(group).size }
            .toMutableMap()
        val lockedPlayerIds = lockedPlayersByGroup.values.flatten().map { it.id }.toSet()
        val playerOrder = fieldPlayers.sortedWith(
            compareBy<LineupPlayer>(
                { candidate ->
                    listOf(PositionGroup.DEFENSE, PositionGroup.LR_MID, PositionGroup.CM_STRIKER)
                        .count { group ->
                            remaining.getOrDefault(group, 0) > 0 && priorHalfGroups[candidate.id] != group
                        }
                },
                { historyByPlayer[it.id]?.minutesPlayed ?: 0.0 },
                { it.name },
            ),
        )

        playerOrder
            .filterNot { it.id in lockedPlayerIds }
            .forEach { player ->
            val allowedGroups = listOf(PositionGroup.DEFENSE, PositionGroup.LR_MID, PositionGroup.CM_STRIKER)
                .filter { group ->
                    remaining.getOrDefault(group, 0) > 0 && priorHalfGroups[player.id] != group
                }
                .ifEmpty {
                    listOf(PositionGroup.DEFENSE, PositionGroup.LR_MID, PositionGroup.CM_STRIKER)
                        .filter { remaining.getOrDefault(it, 0) > 0 }
                }

            val selectedGroup = allowedGroups.minWithOrNull(
                compareBy<PositionGroup>(
                    { historyByPlayer[player.id]?.groupCounts?.get(it) ?: 0 },
                    { result.getValue(it).size },
                    { it.ordinal },
                ),
            ) ?: PositionGroup.DEFENSE

            result.getValue(selectedGroup) += player
            remaining[selectedGroup] = remaining.getOrDefault(selectedGroup, 1) - 1
        }

        return result
    }

    private fun sanitizeManualLocks(
        template: GameTemplateConfig,
        manualGroupLocks: List<ManualGroupLock>,
        playerById: Map<String, LineupPlayer>,
    ): SanitizedLocks {
        val warnings = mutableListOf<String>()
        val locksByHalfGroup = mutableMapOf<Int, MutableMap<PositionGroup, List<LineupPlayer>>>()

        manualGroupLocks
            .groupBy { it.halfNumber }
            .forEach { (halfNumber, halfLocks) ->
                if (halfNumber !in 1..template.halfCount) {
                    warnings += "Ignored manual locks for half $halfNumber because that half does not exist."
                    return@forEach
                }
                val takenPlayerIds = mutableSetOf<String>()
                val groupsForHalf = mutableMapOf<PositionGroup, List<LineupPlayer>>()

                PositionGroup.entries.forEach { group ->
                    val requestedIds = halfLocks
                        .filter { it.positionGroup == group }
                        .flatMap { it.playerIds }
                        .distinct()

                    val selectedPlayers = mutableListOf<LineupPlayer>()
                    requestedIds.forEach { playerId ->
                        val player = playerById[playerId]
                        if (player == null) {
                            warnings += "Ignored a manual lock for an unavailable player in half $halfNumber ${group.label}."
                            return@forEach
                        }
                        if (!takenPlayerIds.add(playerId)) {
                            warnings += "Ignored duplicate manual lock for ${player.name} in half $halfNumber."
                            return@forEach
                        }
                        selectedPlayers += player
                    }

                    val normalizedPlayers = if (group == PositionGroup.GOALIE && selectedPlayers.size > 1) {
                        warnings += "Half $halfNumber goalie lock only supports one player. Keeping ${selectedPlayers.first().name}."
                        selectedPlayers.take(1)
                    } else {
                        selectedPlayers
                    }

                    if (normalizedPlayers.isNotEmpty()) {
                        groupsForHalf[group] = normalizedPlayers
                    }
                }

                locksByHalfGroup[halfNumber] = groupsForHalf
            }

        return SanitizedLocks(
            locksByHalfGroup = locksByHalfGroup,
            warnings = warnings,
        )
    }

    private data class SanitizedLocks(
        val locksByHalfGroup: Map<Int, Map<PositionGroup, List<LineupPlayer>>>,
        val warnings: List<String>,
    )

    private fun generateGroupAssignments(
        halfNumber: Int,
        roundsPerHalf: Int,
        positions: List<FieldPosition>,
        players: List<LineupPlayer>,
        historyByPlayer: Map<String, PlayerSeasonHistory>,
    ): List<GeneratedAssignment> {
        if (players.isEmpty()) return emptyList()
        val slotCounts = players.associate { it.id to 0 }.toMutableMap()
        val positionCounts = mutableMapOf<Pair<String, FieldPosition>, Int>()
        val assignments = mutableListOf<GeneratedAssignment>()
        var previousAssignmentsByPosition = emptyMap<FieldPosition, String>()

        for (roundIndex in 1..roundsPerHalf) {
            val selectedPlayers = players
                .sortedWith(
                    compareBy<LineupPlayer>(
                        { slotCounts[it.id] ?: 0 },
                        { historyByPlayer[it.id]?.minutesPlayed ?: 0.0 },
                        { it.name },
                    ),
                )
                .take(positions.size.coerceAtMost(players.size))
                .toMutableList()

            selectedPlayers.forEach { player ->
                slotCounts[player.id] = slotCounts.getOrDefault(player.id, 0) + 1
            }

            val remainingPlayers = selectedPlayers.toMutableList()
            val currentAssignmentsByPosition = mutableMapOf<FieldPosition, String>()

            // If a player is staying on the field between rounds, keep them in the same exact
            // position and only use the open positions for the actual subs coming in.
            previousAssignmentsByPosition.forEach { (position, playerId) ->
                if (position !in positions) return@forEach
                val stayingPlayer = remainingPlayers.firstOrNull { it.id == playerId } ?: return@forEach
                currentAssignmentsByPosition[position] = stayingPlayer.id
                remainingPlayers.remove(stayingPlayer)
            }

            positions
                .filterNot { it in currentAssignmentsByPosition }
                .forEach { position ->
                    val player = remainingPlayers.minWithOrNull(
                    compareBy<LineupPlayer>(
                        {
                            (historyByPlayer[it.id]?.positionCounts?.get(position) ?: 0) +
                                positionCounts.getOrDefault(it.id to position, 0)
                        },
                        { slotCounts[it.id] ?: 0 },
                        { it.name },
                    ),
                ) ?: return@forEach

                    currentAssignmentsByPosition[position] = player.id
                    remainingPlayers.remove(player)
                }

            positions.forEach { position ->
                val playerId = currentAssignmentsByPosition[position] ?: return@forEach
                positionCounts[playerId to position] = positionCounts.getOrDefault(playerId to position, 0) + 1
                assignments += GeneratedAssignment(
                    halfNumber = halfNumber,
                    roundIndex = roundIndex,
                    playerId = playerId,
                    position = position,
                    positionGroup = position.group,
                )
            }
            previousAssignmentsByPosition = positions.associateWith { position ->
                currentAssignmentsByPosition.getValue(position)
            }
        }

        return assignments
    }
}

File: cloud/shared/lineup-generator.ts
export type WebFormationType = "CLASSIC_U9" | "ATTACK_BACK_THREE";

export interface WebLineupPlayer {
  playerId: string;
  name: string;
  keeperEligible: boolean;
  availableFirstHalf: boolean;
  availableSecondHalf: boolean;
}

export interface WebPlayerHistory {
  keeperAssignments: number;
  minutesPlayed?: number;
  groupCounts: Record<string, number>;
  positionCounts: Record<string, number>;
  totalAssignments: number;
}

export interface WebGeneratedAssignment {
  playerId: string;
  halfNumber: number;
  roundIndex: number;
  position: string;
  positionGroup: string;
}

export interface WebManualLock {
  halfNumber: number;
  positionGroup: string;
  playerIds: string[];
  lockedPosition?: string | null;
}

export interface WebLineupGenerationResult {
  assignments: WebGeneratedAssignment[];
  warnings: string[];
  roundsPerHalf: number;
}

export interface WebLineupGenerationInput {
  players: WebLineupPlayer[];
  formationType: WebFormationType;
  halfDurationMinutes: number;
  substitutionWindowMinutes: number;
  plannedRoundsPerHalf?: number;
  manualLocks?: WebManualLock[];
  historyByPlayer?: Record<string, WebPlayerHistory>;
  variationSeed?: number;
}

export const FORMATIONS = {
  CLASSIC_U9: {
    label: "Classic U9",
    description: "Defense / L-R Mid / CM-Striker",
    positions: ["LEFT_DEFENSE", "RIGHT_DEFENSE", "LEFT_MIDFIELDER", "RIGHT_MIDFIELDER", "CENTER_MIDFIELDER", "STRIKER", "GOALIE"],
    groups: [
      { key: "DEFENSE", positions: ["LEFT_DEFENSE", "RIGHT_DEFENSE"] },
      { key: "LR_MID", positions: ["LEFT_MIDFIELDER", "RIGHT_MIDFIELDER"] },
      { key: "CM_STRIKER", positions: ["CENTER_MIDFIELDER", "STRIKER"] },
    ],
  },
  ATTACK_BACK_THREE: {
    label: "Attack + Back Three",
    description: "Attack / Defense / Goalie",
    positions: ["STRIKER", "LEFT_MIDFIELDER", "RIGHT_MIDFIELDER", "LEFT_DEFENSE", "CENTER_DEFENSE", "RIGHT_DEFENSE", "GOALIE"],
    groups: [
      { key: "ATTACK", positions: ["STRIKER", "LEFT_MIDFIELDER", "RIGHT_MIDFIELDER"] },
      { key: "DEFENSE", positions: ["LEFT_DEFENSE", "CENTER_DEFENSE", "RIGHT_DEFENSE"] },
    ],
  },
} as const;

export function generateWebLineup(input: WebLineupGenerationInput): WebLineupGenerationResult {
  const formation = FORMATIONS[input.formationType];
  const suggestedRounds = Math.ceil(Math.max(input.halfDurationMinutes, 1) / Math.max(input.substitutionWindowMinutes, 1)) + 1;
  const roundsPerHalf = Math.max(1, Math.round(input.plannedRoundsPerHalf ?? suggestedRounds));
  const history = input.historyByPlayer ?? {};
  const seed = input.variationSeed ?? 0;
  const sanitizedLocks = sanitizeManualLocks(input, formation);
  const warnings = [...sanitizedLocks.warnings];
  const assignments: WebGeneratedAssignment[] = [];
  const firstHalfGroups = new Map<string, string>();
  const selectedKeepers: WebLineupPlayer[] = [];

  for (const halfNumber of [1, 2]) {
    const available = input.players.filter(player => halfNumber === 1 ? player.availableFirstHalf : player.availableSecondHalf);
    if (available.length < formation.positions.length) {
      warnings.push(`Half ${halfNumber} needs at least ${formation.positions.length} available players to fill every position.`);
      continue;
    }

    const lockedKeeper = sanitizedLocks.groups.get(halfNumber)?.get("GOALIE")?.[0];
    const keeperPool = available.filter(candidate =>
      selectedKeepers.every(selected => selected.playerId !== candidate.playerId) || available.length <= selectedKeepers.length + 1
    );
    const keeper = lockedKeeper ?? [...(keeperPool.length ? keeperPool : available)].sort((left, right) =>
      compareNumbers(left.keeperEligible ? 0 : 1, right.keeperEligible ? 0 : 1) ||
      compareNumbers(history[left.playerId]?.keeperAssignments ?? 0, history[right.playerId]?.keeperAssignments ?? 0) ||
      compareNumbers(history[left.playerId]?.minutesPlayed ?? 0, history[right.playerId]?.minutesPlayed ?? 0) ||
      compareNumbers(variationRank(seed, `keeper-${halfNumber}-${left.playerId}`), variationRank(seed, `keeper-${halfNumber}-${right.playerId}`)) ||
      left.name.localeCompare(right.name)
    )[0];
    selectedKeepers.push(keeper);

    const fieldPlayers = available.filter(player => player.playerId !== keeper.playerId);
    const locksForHalf = sanitizedLocks.groups.get(halfNumber) ?? new Map();
    const positionLocksForHalf: Map<string, WebLineupPlayer[]> = sanitizedLocks.positions.get(halfNumber) ?? new Map();
    const lockedPlayersByGroup = new Map<string, WebLineupPlayer[]>();
    for (const group of formation.groups) {
      const positionLocked = group.positions.flatMap(position => positionLocksForHalf.get(position) ?? []);
      lockedPlayersByGroup.set(group.key, uniquePlayers([...(locksForHalf.get(group.key) ?? []), ...positionLocked]).filter(player => player.playerId !== keeper.playerId));
    }
    const capacities = createGroupCapacities(
      formation.groups.map(group => group.key),
      fieldPlayers.length,
      new Map([...lockedPlayersByGroup].map(([group, players]) => [group, players.length])),
    );
    if ([...lockedPlayersByGroup.values()].flat().length > fieldPlayers.length) warnings.push(`Half ${halfNumber} has more locked field players than available field spots.`);
    if (capacities.some(capacity => capacity < 3)) {
      warnings.push(`Half ${halfNumber} cannot keep every field group at three players with the current availability.`);
    }
    const priorGroups = halfNumber === 2 ? firstHalfGroups : new Map<string, string>();
    if (halfNumber === 2) {
      for (const [group, lockedPlayers] of lockedPlayersByGroup) {
        for (const player of lockedPlayers) {
          if (priorGroups.get(player.playerId) === group) warnings.push(`${player.name} is manually locked into ${groupLabel(group)} for both halves.`);
        }
      }
    }
    const grouped = assignFieldGroups(
      formation.groups.map(group => group.key),
      fieldPlayers,
      capacities,
      lockedPlayersByGroup,
      priorGroups,
      history,
    );
    if (halfNumber === 1) {
      for (const [group, groupedPlayers] of grouped) for (const player of groupedPlayers) firstHalfGroups.set(player.playerId, group);
    }

    for (let roundIndex = 1; roundIndex <= roundsPerHalf; roundIndex += 1) {
      assignments.push({ playerId: keeper.playerId, halfNumber, roundIndex, position: "GOALIE", positionGroup: "GOALIE" });
    }

    for (const group of formation.groups) {
      assignments.push(...generateGroupRounds({
        halfNumber,
        roundsPerHalf,
        groupKey: group.key,
        positions: [...group.positions],
        players: grouped.get(group.key) ?? [],
        lockedPlayerIdsByPosition: Object.fromEntries(
          group.positions.map(position => [
            position,
            (positionLocksForHalf.get(position) ?? []).map(player => player.playerId),
          ]),
        ),
        history,
        seed,
      }));
    }
  }

  const positionOrder = new Map<string, number>(formation.positions.map((position, index) => [position, index]));
  assignments.sort((left, right) =>
    compareNumbers(left.halfNumber, right.halfNumber) ||
    compareNumbers(left.roundIndex, right.roundIndex) ||
    compareNumbers(positionOrder.get(left.position) ?? 999, positionOrder.get(right.position) ?? 999)
  );
  return { assignments, warnings: [...new Set(warnings)], roundsPerHalf };
}

function generateGroupRounds(input: {
  halfNumber: number;
  roundsPerHalf: number;
  groupKey: string;
  positions: string[];
  players: WebLineupPlayer[];
  lockedPlayerIdsByPosition: Record<string, string[]>;
  history: Record<string, WebPlayerHistory>;
  seed: number;
}): WebGeneratedAssignment[] {
  const output: WebGeneratedAssignment[] = [];
  const appearances = new Map(input.players.map(player => [player.playerId, 0]));
  const positionCounts = new Map<string, number>();
  let previous = new Map<string, string>();

  for (let roundIndex = 1; roundIndex <= input.roundsPerHalf; roundIndex += 1) {
    const selected = [...input.players].sort((left, right) =>
      compareNumbers(appearances.get(left.playerId) ?? 0, appearances.get(right.playerId) ?? 0) ||
      compareNumbers(input.history[left.playerId]?.minutesPlayed ?? 0, input.history[right.playerId]?.minutesPlayed ?? 0) ||
      compareNumbers(variationRank(input.seed, `round-${input.halfNumber}-${roundIndex}-${left.playerId}`), variationRank(input.seed, `round-${input.halfNumber}-${roundIndex}-${right.playerId}`)) ||
      left.name.localeCompare(right.name)
    ).slice(0, input.positions.length);
    for (const position of input.positions) {
      const lockedCandidates = input.players
        .filter(player => input.lockedPlayerIdsByPosition[position]?.includes(player.playerId))
        .sort((left, right) =>
          compareNumbers(appearances.get(left.playerId) ?? 0, appearances.get(right.playerId) ?? 0) ||
          compareNumbers(positionCounts.get(`${left.playerId}:${position}`) ?? 0, positionCounts.get(`${right.playerId}:${position}`) ?? 0) ||
          compareNumbers(variationRank(input.seed, `locked-position-${input.halfNumber}-${roundIndex}-${position}-${left.playerId}`), variationRank(input.seed, `locked-position-${input.halfNumber}-${roundIndex}-${position}-${right.playerId}`)) ||
          left.name.localeCompare(right.name)
        );
      const lockedPlayer = lockedCandidates[0];
      if (!lockedPlayer || selected.some(player => player.playerId === lockedPlayer.playerId)) continue;
      const replaceable = selected
        .filter(player => !Object.values(input.lockedPlayerIdsByPosition).some(ids => ids.includes(player.playerId)))
        .sort((left, right) =>
          compareNumbers(appearances.get(right.playerId) ?? 0, appearances.get(left.playerId) ?? 0) || right.name.localeCompare(left.name)
        )[0] ?? [...selected].sort((left, right) =>
          compareNumbers(appearances.get(right.playerId) ?? 0, appearances.get(left.playerId) ?? 0) || right.name.localeCompare(left.name)
        )[0];
      const replacementIndex = replaceable ? selected.findIndex(player => player.playerId === replaceable.playerId) : -1;
      if (replacementIndex >= 0) selected[replacementIndex] = lockedPlayer;
    }
    selected.forEach(player => appearances.set(player.playerId, (appearances.get(player.playerId) ?? 0) + 1));

    const remaining = new Map(selected.map(player => [player.playerId, player]));
    const current = new Map<string, string>();
    for (const position of input.positions) {
      const lockedPlayer = [...remaining.values()]
        .filter(player => input.lockedPlayerIdsByPosition[position]?.includes(player.playerId))
        .sort((left, right) =>
          compareNumbers(positionCounts.get(`${left.playerId}:${position}`) ?? 0, positionCounts.get(`${right.playerId}:${position}`) ?? 0) ||
          compareNumbers(appearances.get(left.playerId) ?? 0, appearances.get(right.playerId) ?? 0) ||
          compareNumbers(variationRank(input.seed, `locked-assignment-${input.halfNumber}-${roundIndex}-${position}-${left.playerId}`), variationRank(input.seed, `locked-assignment-${input.halfNumber}-${roundIndex}-${position}-${right.playerId}`)) ||
          left.name.localeCompare(right.name)
        )[0];
      if (lockedPlayer) {
        current.set(position, lockedPlayer.playerId);
        remaining.delete(lockedPlayer.playerId);
      }
    }
    for (const position of input.positions) {
      if (current.has(position)) continue;
      const priorPlayerId = previous.get(position);
      if (priorPlayerId && remaining.has(priorPlayerId)) {
        current.set(position, priorPlayerId);
        remaining.delete(priorPlayerId);
      }
    }
    for (const position of input.positions.filter(position => !current.has(position))) {
      const player = [...remaining.values()].sort((left, right) =>
        compareNumbers(
          (input.history[left.playerId]?.positionCounts?.[position] ?? 0) + (positionCounts.get(`${left.playerId}:${position}`) ?? 0),
          (input.history[right.playerId]?.positionCounts?.[position] ?? 0) + (positionCounts.get(`${right.playerId}:${position}`) ?? 0),
        ) ||
        compareNumbers(appearances.get(left.playerId) ?? 0, appearances.get(right.playerId) ?? 0) ||
        compareNumbers(variationRank(input.seed, `position-${input.halfNumber}-${roundIndex}-${position}-${left.playerId}`), variationRank(input.seed, `position-${input.halfNumber}-${roundIndex}-${position}-${right.playerId}`)) ||
        left.name.localeCompare(right.name)
      )[0];
      if (!player) continue;
      current.set(position, player.playerId);
      remaining.delete(player.playerId);
    }
    for (const position of input.positions) {
      const playerId = current.get(position);
      if (!playerId) continue;
      const key = `${playerId}:${position}`;
      positionCounts.set(key, (positionCounts.get(key) ?? 0) + 1);
      output.push({ playerId, halfNumber: input.halfNumber, roundIndex, position, positionGroup: input.groupKey });
    }
    previous = current;
  }
  return output;
}

function createGroupCapacities(
  groupKeys: string[],
  fieldPlayerCount: number,
  lockedCounts: Map<string, number>,
): number[] {
  const base = Math.floor(fieldPlayerCount / groupKeys.length);
  const extra = fieldPlayerCount % groupKeys.length;
  const capacities = groupKeys.map((_, index) => base + (index < extra ? 1 : 0));
  groupKeys.forEach((group, groupIndex) => {
    let deficit = (lockedCounts.get(group) ?? 0) - capacities[groupIndex];
    if (deficit <= 0) return;
    capacities[groupIndex] = lockedCounts.get(group) ?? capacities[groupIndex];
    const donors = groupKeys.map((donor, index) => ({ donor, index }))
      .filter(item => item.index !== groupIndex)
      .sort((left, right) =>
        (capacities[right.index] - (lockedCounts.get(right.donor) ?? 0)) -
        (capacities[left.index] - (lockedCounts.get(left.donor) ?? 0))
      );
    for (const donor of donors) {
      while (deficit > 0 && capacities[donor.index] > (lockedCounts.get(donor.donor) ?? 0)) {
        capacities[donor.index] -= 1;
        deficit -= 1;
      }
    }
  });
  return capacities;
}

function assignFieldGroups(
  groupKeys: string[],
  fieldPlayers: WebLineupPlayer[],
  capacities: number[],
  lockedPlayersByGroup: Map<string, WebLineupPlayer[]>,
  priorGroups: Map<string, string>,
  history: Record<string, WebPlayerHistory>,
): Map<string, WebLineupPlayer[]> {
  const result = new Map(groupKeys.map(group => [group, [...(lockedPlayersByGroup.get(group) ?? [])]]));
  const remaining = new Map(groupKeys.map((group, index) => [group, capacities[index] - (result.get(group)?.length ?? 0)]));
  const lockedIds = new Set([...lockedPlayersByGroup.values()].flat().map(player => player.playerId));
  const playerOrder = [...fieldPlayers].sort((left, right) => {
    const leftOptions = groupKeys.filter(group => (remaining.get(group) ?? 0) > 0 && priorGroups.get(left.playerId) !== group).length;
    const rightOptions = groupKeys.filter(group => (remaining.get(group) ?? 0) > 0 && priorGroups.get(right.playerId) !== group).length;
    return compareNumbers(leftOptions, rightOptions) ||
      compareNumbers(history[left.playerId]?.minutesPlayed ?? 0, history[right.playerId]?.minutesPlayed ?? 0) ||
      left.name.localeCompare(right.name);
  });
  for (const player of playerOrder.filter(player => !lockedIds.has(player.playerId))) {
    let allowed = groupKeys.filter(group => (remaining.get(group) ?? 0) > 0 && priorGroups.get(player.playerId) !== group);
    if (!allowed.length) allowed = groupKeys.filter(group => (remaining.get(group) ?? 0) > 0);
    const selected = [...allowed].sort((left, right) =>
      compareNumbers(history[player.playerId]?.groupCounts?.[left] ?? 0, history[player.playerId]?.groupCounts?.[right] ?? 0) ||
      compareNumbers(result.get(left)?.length ?? 0, result.get(right)?.length ?? 0) ||
      compareNumbers(groupKeys.indexOf(left), groupKeys.indexOf(right))
    )[0] ?? groupKeys[0];
    result.get(selected)?.push(player);
    remaining.set(selected, (remaining.get(selected) ?? 1) - 1);
  }
  return result;
}

function sanitizeManualLocks(
  input: WebLineupGenerationInput,
  formation: (typeof FORMATIONS)[WebFormationType],
): { groups: Map<number, Map<string, WebLineupPlayer[]>>; positions: Map<number, Map<string, WebLineupPlayer[]>>; warnings: string[] } {
  const warnings: string[] = [];
  const groups = new Map<number, Map<string, WebLineupPlayer[]>>();
  const positions = new Map<number, Map<string, WebLineupPlayer[]>>();
  const playerById = new Map(input.players.map(player => [player.playerId, player]));
  const activeGroups = new Set([...formation.groups.map(group => group.key as string), "GOALIE"]);
  const activePositions = new Set(formation.positions as readonly string[]);
  const groupedLocks = new Map<number, WebManualLock[]>();
  for (const lock of input.manualLocks ?? []) groupedLocks.set(lock.halfNumber, [...(groupedLocks.get(lock.halfNumber) ?? []), lock]);
  for (const [halfNumber, halfLocks] of groupedLocks) {
    if (halfNumber < 1 || halfNumber > 2) {
      warnings.push(`Ignored manual locks for half ${halfNumber} because that half does not exist.`);
      continue;
    }
    const taken = new Set<string>();
    const halfGroups = new Map<string, WebLineupPlayer[]>();
    const halfPositions = new Map<string, WebLineupPlayer[]>();
    for (const position of formation.positions) {
      const requested = [...new Set(halfLocks.filter(lock => lock.lockedPosition === position).flatMap(lock => lock.playerIds))];
      const selected = sanitizeRequestedPlayers(requested, halfNumber, label(position), playerById, taken, warnings);
      if (selected.length) halfPositions.set(position, selected);
    }
    for (const group of [...formation.groups.map(item => item.key as string), "GOALIE"]) {
      const requested = [...new Set(halfLocks.filter(lock => lock.positionGroup === group && !lock.lockedPosition).flatMap(lock => lock.playerIds))];
      let selected = sanitizeRequestedPlayers(requested, halfNumber, groupLabel(group), playerById, taken, warnings);
      if (group === "GOALIE" && selected.length > 1) {
        warnings.push(`Half ${halfNumber} goalie lock only supports one player. Keeping ${selected[0].name}.`);
        selected = selected.slice(0, 1);
      }
      if (selected.length) halfGroups.set(group, selected);
    }
    for (const lock of halfLocks.filter(lock => lock.lockedPosition && !activePositions.has(lock.lockedPosition) && lock.playerIds.length)) {
      warnings.push(`Ignored ${label(lock.lockedPosition!)} locks because that position is not used by ${formation.label}.`);
    }
    for (const lock of halfLocks.filter(lock => !lock.lockedPosition && !activeGroups.has(lock.positionGroup) && lock.playerIds.length)) {
      warnings.push(`Ignored ${groupLabel(lock.positionGroup)} locks because that group is not used by ${formation.label}.`);
    }
    groups.set(halfNumber, halfGroups);
    positions.set(halfNumber, halfPositions);
  }
  return { groups, positions, warnings };
}

function sanitizeRequestedPlayers(
  requested: string[],
  halfNumber: number,
  lockLabel: string,
  playerById: Map<string, WebLineupPlayer>,
  taken: Set<string>,
  warnings: string[],
): WebLineupPlayer[] {
  const selected: WebLineupPlayer[] = [];
  for (const playerId of requested) {
    const player = playerById.get(playerId);
    if (!player) {
      warnings.push(`Ignored a manual lock for an unavailable player in half ${halfNumber} ${lockLabel}.`);
      continue;
    }
    const available = halfNumber === 1 ? player.availableFirstHalf : player.availableSecondHalf;
    if (!available) {
      warnings.push(`Ignored a manual lock for ${player.name} in half ${halfNumber} because that player is unavailable that half.`);
      continue;
    }
    if (taken.has(playerId)) {
      warnings.push(`Ignored duplicate manual lock for ${player.name} in half ${halfNumber}.`);
      continue;
    }
    taken.add(playerId);
    selected.push(player);
  }
  return selected;
}

function uniquePlayers(players: WebLineupPlayer[]): WebLineupPlayer[] {
  return [...new Map(players.map(player => [player.playerId, player])).values()];
}

function groupLabel(group: string): string {
  return ({ ATTACK: "Attack", DEFENSE: "Defense", LR_MID: "L/R Mid", CM_STRIKER: "CM/Striker", GOALIE: "Goalie" } as Record<string, string>)[group] ?? group;
}

function label(value: string): string {
  return value.toLowerCase().split("_").map(part => part[0]?.toUpperCase() + part.slice(1)).join(" ");
}

function compareNumbers(left: number, right: number): number {
  return left - right;
}

function kotlinHash(value: string): number {
  let hash = 0;
  for (let index = 0; index < value.length; index += 1) hash = ((hash << 5) - hash + value.charCodeAt(index)) | 0;
  return hash;
}

function variationRank(seed: number, key: string): number {
  return kotlinHash(`${seed}${key}`);
}
